tests/unit/server/queue_consumer_tests.py

```diff
@@ -0,0 +1,226 @@
+    @pytest.fixture
+    def build_server(self):
+        server = [None]
+
+        def _build_server(max_concurrency=5, pump_raises=None, handler_raises=None):
+            server[0] = QueueConsumerServer.new(
+                consumer_factory=FakeQueueConsumerFactory(
+                    pump_raises=pump_raises, handler_raises=handler_raises
+                ),
+                max_concurrency=max_concurrency,
+                listener=mock.Mock(spec=socket.socket),
+            )
+            return server[0]
+
+        yield _build_server
+
+        try:
+            if server[0] is not None:
+                server[0].stop()
+        except AssertionError:
+            pass
+
+    @pytest.fixture
+    def server(self, build_server):
+        return build_server()
+
+    def test_new(self):
+        max_concurrency = 5
+        server = QueueConsumerServer.new(
+            consumer_factory=FakeQueueConsumerFactory(),
+            max_concurrency=max_concurrency,
+            listener=mock.Mock(spec=socket.socket),
+        )
+        assert not server.started
+        assert not server.stopped
+        assert not server.pump.started
+        assert not server.pump.stopped
+        assert server.pump.work_queue.maxsize == 7
+        server.healthcheck_server.start.assert_not_called()
+        server.healthcheck_server.stop.assert_not_called()
+        assert len(server.handlers) == max_concurrency
+        assert len(server.handlers) == len(server.threads)
+        for handler in server.handlers:
+            assert not handler.started
+            assert not handler.stopped
+
+    def test_start(self, server):
+        server.start()
+        assert server.started
+        assert not server.stopped
+        assert server.pump.started
+        assert not server.pump.stopped
+        server.healthcheck_server.start.assert_called_once()
+        server.healthcheck_server.stop.assert_not_called()
+        for handler in server.handlers:
+            assert handler.started
+            assert not handler.stopped
+        with pytest.raises(AssertionError):
+            server.start()
+
+    def test_stop(self, server):
+        server.start()
+        server.stop()
+        assert server.started
+        assert server.stopped
+        assert server.pump.started
+        assert server.pump.stopped
+        server.healthcheck_server.start.assert_called_once()
+        server.healthcheck_server.stop.assert_called_once()
+        for handler in server.handlers:
+            assert handler.started
+            assert handler.stopped
+        with pytest.raises(AssertionError):
+            server.stop()
+
+    def test_stop_before_start(self, server):
+        with pytest.raises(AssertionError):
+            server.stop()
+
+    def test_pump_exception_terminates(self, build_server):
+        server = build_server(max_concurrency=1, pump_raises=Exception())
+        server._terminate = mock.Mock()
+        server.start()
+        time.sleep(2)
```

Comment:
oof, this'll add a chunk o' time to the tests. any way to do this with shorter sleeps?